### ISSUE TITLE
f16 regressions in the LLVM16 branch on macOS

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#15002]
On my x86_64-macos laptop, I can reproduce the issue this way:

```zig
const std = @import("std");

pub fn main() !void {
    foo(4.0);
}

fn foo(y: f16) void {
    std.debug.print("y={d}\n", .{y});
}
```

Output:

```
y=0
```

The LLVM IR emitted has `half 0xH0000` at the callsite. The compiler has incorrectly converted 4.0 to 0.

I'm not sure why this is happening but it might be related to these [release notes](https://releases.llvm.org/16.0.0/docs/ReleaseNotes.html#changes-to-the-x86-backend):

> Support half type on SSE2 and above targets following X86 psABI.
>
> Support rdpru instruction on Zen2 and above targets.
>
> During this release, half type has an ABI breaking change to provide the support for the ABI of _Float16 type on SSE2 and above following X86 psABI. ([D107082](https://reviews.llvm.org/D107082))
>
> The change may affect the current use of half includes (but is not limited to):
>
> * Frontends generating half type in function passing and/or returning arguments.
> * Downstream runtimes providing any half conversion builtins assuming the old ABI.
> * Projects built with LLVM 15.0 but using early versions of compiler-rt.
>
> When you find failures with half type, check the calling conversion of the code and switch it to the new ABI.

The issue manifests on the CI in this way:

```
zig test Debug native: error: thread 255518 panic: reached unreachable code
Analyzing /Users/runner/work/zig/zig/test/behavior/cast.zig: behavior/cast.zig:test.comptime_int @intToFloat
      %489 = float(1234)
    > %490 = cmp_eq(%470, %489) 
      %491 = break_inline(%488, %490)
    For full context, use the command
      zig ast-check -t /Users/runner/work/zig/zig/test/behavior/cast.zig

  in /Users/runner/work/zig/zig/test/behavior/cast.zig: behavior/cast.zig:test.comptime_int @intToFloat
    > %488 = call(.auto, %486, [
        {%489..%491},
      ]) 
  in /Users/runner/work/zig/zig/test/behavior/cast.zig: behavior/cast.zig:test.comptime_int @intToFloat
    > %466 = block({%467..%499}) 
  in /Users/runner/work/zig/zig/test/behavior/cast.zig: behavior/cast.zig:test.comptime_int @intToFloat
    > %463 = block({%464..%642}) 

/Users/runner/work/zig/zig/lib/std/math.zig:1465:9: 0x1015ceb5a in order__anon_155215 (zig)
        unreachable;
        ^
/Users/runner/work/zig/zig/src/value.zig:2065:111: 0x1015ccf5f in orderAdvanced (zig)
                    .float_16 => return std.math.order(lhs.castTag(.float_16).?.data, rhs.castTag(.float_16).?.data),
                                                                                                              ^
/Users/runner/work/zig/zig/src/value.zig:2121:34: 0x101944c17 in compareHeteroAdvanced (zig)
        return (try orderAdvanced(lhs, rhs, target, opt_sema)).compare(op);
                                 ^
/Users/runner/work/zig/zig/src/Sema.zig:29060:52: 0x10192cd49 in cmpNumeric (zig)
                if (try Value.compareHeteroAdvanced(lhs_val, op, rhs_val, target, sema)) {
                                                   ^
/Users/runner/work/zig/zig/src/Sema.zig:15132:31: 0x10151fa6f in analyzeCmp (zig)
        return sema.cmpNumeric(block, src, lhs, rhs, op, lhs_src, rhs_src);
                              ^
/Users/runner/work/zig/zig/src/Sema.zig:15049:27: 0x100f4d732 in zirCmpEq (zig)
    return sema.analyzeCmp(block, src, lhs, rhs, op, lhs_src, rhs_src, true);
                          ^
/Users/runner/work/zig/zig/src/Sema.zig:917:115: 0x100b30ecd in analyzeBodyInner (zig)
            .cmp_eq                       => try sema.zirCmpEq(block, inst, .eq, Air.Inst.Tag.fromCmpOp(.eq, block.float_mode == .Optimized)),
                                                                                                                  ^
/Users/runner/work/zig/zig/src/Sema.zig:812:45: 0x1008fa9b6 in analyzeBodyBreak (zig)
    const break_inst = sema.analyzeBodyInner(block, body) catch |err| switch (err) {
                                            ^
/Users/runner/work/zig/zig/src/Sema.zig:762:50: 0x10150fad4 in resolveBody (zig)
    const break_data = (try sema.analyzeBodyBreak(block, body)) orelse
                                                 ^
/Users/runner/work/zig/zig/src/Sema.zig:6272:63: 0x100f4a15d in zirCall (zig)
        const resolved = try sema.resolveBody(block, args_body[arg_start..arg_end], inst);
                                                              ^
/Users/runner/work/zig/zig/src/Sema.zig:913:62: 0x100b30b14 in analyzeBodyInner (zig)
            .call                         => try sema.zirCall(block, inst),
                                                             ^
/Users/runner/work/zig/zig/src/Sema.zig:5429:34: 0x10155831a in resolveBlockBody (zig)
        if (sema.analyzeBodyInner(child_block, body)) |_| {
                                 ^
/Users/runner/work/zig/zig/src/Sema.zig:5412:85: 0x101013bad in zirBlock (zig)
    return sema.resolveBlockBody(parent_block, src, &child_block, body, inst, &label.merges);
                                                                                    ^
/Users/runner/work/zig/zig/src/Sema.zig:1454:69: 0x100b40d11 in analyzeBodyInner (zig)
                if (!block.is_comptime) break :blk try sema.zirBlock(block, inst);
                                                                    ^
/Users/runner/work/zig/zig/src/Sema.zig:5429:34: 0x10155831a in resolveBlockBody (zig)
        if (sema.analyzeBodyInner(child_block, body)) |_| {
                                 ^
/Users/runner/work/zig/zig/src/Sema.zig:5412:85: 0x101013bad in zirBlock (zig)
    return sema.resolveBlockBody(parent_block, src, &child_block, body, inst, &label.merges);
                                                                                    ^
/Users/runner/work/zig/zig/src/Sema.zig:1454:69: 0x100b40d11 in analyzeBodyInner (zig)
                if (!block.is_comptime) break :blk try sema.zirBlock(block, inst);
                                                                    ^
/Users/runner/work/zig/zig/src/Sema.zig:795:30: 0x100ef1bb9 in analyzeBody (zig)
    _ = sema.analyzeBodyInner(block, body) catch |err| switch (err) {
                             ^
/Users/runner/work/zig/zig/src/Module.zig:5546:43: 0x100b14cc7 in analyzeFnBody (zig)
    sema.analyzeBody(&inner_block, fn_info.body) catch |err| switch (err) {
                                          ^
/Users/runner/work/zig/zig/src/Module.zig:4244:40: 0x1008d8b26 in ensureFuncBodyAnalyzed (zig)
            var air = mod.analyzeFnBody(func, sema_arena) catch |err| switch (err) {
                                       ^
/Users/runner/work/zig/zig/src/Compilation.zig:3116:42: 0x1008d68c9 in processOneJob (zig)
            module.ensureFuncBodyAnalyzed(func) catch |err| switch (err) {
                                         ^
/Users/runner/work/zig/zig/src/Compilation.zig:3054:30: 0x100776ee4 in performAllTheWork (zig)
            try processOneJob(comp, work_item, main_progress_node);
                             ^
/Users/runner/work/zig/zig/src/Compilation.zig:2011:31: 0x10077344d in update (zig)
    try comp.performAllTheWork(main_progress_node);
                              ^
/Users/runner/work/zig/zig/src/main.zig:3502:36: 0x1007a3757 in serve (zig)
                    try comp.update(main_progress_node);
                                   ^
/Users/runner/work/zig/zig/src/main.zig:3305:31: 0x10062ed99 in buildOutputType (zig)
                test_exec_args.items,
                              ^
/Users/runner/work/zig/zig/src/main.zig:273:31: 0x1005ff4ff in mainArgs (zig)
        return buildOutputType(gpa, arena, args, .zig_test);
                              ^
/Users/runner/work/zig/zig/src/main.zig:211:20: 0x1005fe887 in main (zig)
    return mainArgs(gpa, arena, args);
                   ^
/Users/runner/work/zig/zig/lib/std/start.zig:617:37: 0x100601015 in main (zig)
            const result = root.main() catch |err| {
                                    ^
???:?:?: 0x7fff20408f3c in ??? (???)
???:?:?: 0x13 in ??? (???)

zig test Debug native: error: the following command terminated unexpectedly:
/Users/runner/work/zig/zig/build/stage3/bin/zig test /Users/runner/work/zig/zig/test/behavior.zig --cache-dir /Users/runner/work/zig/zig/build/zig-local-cache --global-cache-dir /Users/runner/work/zig/zig/build/zig-global-cache --name test -fno-single-threaded -I /Users/runner/work/zig/zig/test -L /Users/runner/zig+llvm+lld+clang-x86_64-macos-none-0.11.0-dev.2182+0d4e223ab/lib -I /Users/runner/zig+llvm+lld+clang-x86_64-macos-none-0.11.0-dev.2182+0d4e223ab/include --zig-lib-dir /Users/runner/work/zig/zig/lib --enable-cache --listen=- 
```

**Comment by andrewrk** [comment]
Clue: I tried running the example above on x86_64-macos using a cross-compiled zig binary, and it correctly printed y=4. So this seems like a miscompilation that is making it into the compiler.

**Comment by mikdusan** [comment]
for me, on x86_64 macos 13.2.1, zig built against llvm-16.0.0, produced from xcode clang, (not zig-bootstrap):

```sh
$ /opt/llvm-macos13.0-x86_64-16.0.0-release/bin/llvm-config --version
16.0.0

$ zig version
0.11.0-dev.2158+c31007bb4

$ cat bug0.zig
const std = @import("std");

pub fn main() !void {
    foo(4.0);
}

fn foo(y: f16) void {
    std.debug.print("y={d}\n", .{y});
}

$ zig run bug0.zig
y=4
```

### bug0.ll

```llvm
; ModuleID = 'bug0'
source_filename = "bug0"
target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-macosx13.2.1-unknown"

define internal fastcc i16 @bug0.main(ptr nonnull %0) unnamed_addr #0 !dbg !2660 {
Entry:
  call fastcc void @bug0.foo(half 0xH4400), !dbg !2664
  ret i16 0, !dbg !2664
}

; Function Attrs: nounwind sspstrong
define internal fastcc void @bug0.foo(half %0) unnamed_addr #0 !dbg !2666 {
Entry:
  %StackGuardSlot = alloca ptr, align 8
  %1 = call ptr @llvm.stackguard()
  call void @llvm.stackprotector(ptr %1, ptr %StackGuardSlot)
  %2 = alloca %1, align 2
  %3 = alloca half, align 2
  store half %0, ptr %3, align 2
  call void @llvm.dbg.declare(metadata ptr %3, metadata !2671, metadata !DIExpression()), !dbg !2672
  %4 = getelementptr inbounds %1, ptr %2, i32 0, i32 0, !dbg !2673
  store half %0, ptr %4, align 2, !dbg !2673
  call fastcc void @debug.print__anon_1735(ptr %2), !dbg !2673
  %5 = call ptr @llvm.stackguard()
  %6 = load volatile ptr, ptr %StackGuardSlot, align 8
  %7 = icmp eq ptr %5, %6
  br i1 %7, label %SP_return, label %CallStackCheckFailBlk, !prof !2594

SP_return:                                        ; preds = %Entry
  ret void, !dbg !2673

CallStackCheckFailBlk:                            ; preds = %Entry
  call void @__stack_chk_fail(), !dbg !2675
  unreachable, !dbg !2675
}
```

**Comment by mikdusan** [comment]
```asm
_bug0.main:
Lfunc_begin6:
    .loc    25 3 0 is_stmt 1
    .cfi_startproc
    push    rbp
    .cfi_def_cfa_offset 16
    .cfi_offset rbp, -16
    mov rbp, rsp
    .cfi_def_cfa_register rbp
    vpinsrw xmm0, xmm0, word ptr [rip + LCPI6_0], 0
Ltmp18:
    .loc    25 4 8 prologue_end
    call    _bug0.foo
    xor eax, eax
    pop rbp
    ret
Ltmp19:
Lfunc_end6:
    .cfi_endproc

    .p2align    4, 0x90

_bug0.foo:
Lfunc_begin7:
    .loc    25 7 0
    .cfi_startproc
    push    rbp
    .cfi_def_cfa_offset 16
    .cfi_offset rbp, -16
    mov rbp, rsp
    .cfi_def_cfa_register rbp
    sub rsp, 32
    mov rax, qword ptr [rip + ___stack_chk_guard@GOTPCREL]
    mov rax, qword ptr [rax]
    mov qword ptr [rbp - 24], rax
    mov rax, qword ptr [rip + ___stack_chk_guard@GOTPCREL]
    mov rax, qword ptr [rax]
    mov qword ptr [rbp - 8], rax
    vpextrw word ptr [rbp - 26], xmm0, 0
Ltmp20:
    .loc    25 8 20 prologue_end
    vpextrw word ptr [rbp - 16], xmm0, 0
    lea rdi, [rbp - 16]
    call    _debug.print__anon_1735
    mov rax, qword ptr [rip + ___stack_chk_guard@GOTPCREL]
    mov rax, qword ptr [rax]
    mov rcx, qword ptr [rbp - 8]
    cmp rax, rcx
    jne LBB7_3
    .loc    25 0 20 is_stmt 0
    mov rax, qword ptr [rip + ___stack_chk_guard@GOTPCREL]
    mov rax, qword ptr [rax]
    mov rcx, qword ptr [rbp - 24]
    cmp rax, rcx
    jne LBB7_4
    .loc    25 8 20
    add rsp, 32
    pop rbp
    ret
Ltmp21:
LBB7_3:
    .loc    25 0 0
    call    ___stack_chk_fail
    ud2
LBB7_4:
    call    ___stack_chk_fail
    ud2
Ltmp22:
Lfunc_end7:
    .cfi_endproc
```

**Comment by mikdusan** [comment]
another hint:

I built zig llvm16 branch again, this time against `zig+llvm+lld+clang-x86_64-macos-none-0.11.0-dev.2182+0d4e223ab.tar.xz` for llvm, and the issue manifests again.

**Comment by andrewrk** [comment]
My reproduction is on macOS 11.6.1, built against llvm-16.0.0, produced from xcode clang (not zig-bootstrap). When I use the tarball, the issue does not manifest.

Another piece of data: it looks like LLVM's change of f16 ABI had to do with the presence of certain CPU features. Output of `zig build-obj --show-builtin`:

```
const std = @import("std");
/// Zig version. When writing code that supports multiple versions of Zig, prefer
/// feature detection (i.e. with `@hasDecl` or `@hasField`) over version checks.
pub const zig_version = std.SemanticVersion.parse(zig_version_string) catch unreachable;
pub const zig_version_string = "0.11.0-dev.2191+5c4bbd065";
pub const zig_backend = std.builtin.CompilerBackend.stage2_x86_64;

pub const output_mode = std.builtin.OutputMode.Obj;
pub const link_mode = std.builtin.LinkMode.Static;
pub const is_test = false;
pub const single_threaded = false;
pub const abi = std.Target.Abi.none;
pub const cpu: std.Target.Cpu = .{
    .arch = .x86_64,
    .model = &std.Target.x86.cpu.haswell,
    .features = std.Target.x86.featureSet(&[_]std.Target.x86.Feature{
        .@"64bit",
        .aes,
        .allow_light_256_bit,
        .avx,
        .avx2,
        .bmi,
        .bmi2,
        .cmov,
        .crc32,
        .cx16,
        .cx8,
        .ermsb,
        .f16c,
        .false_deps_lzcnt_tzcnt,
        .false_deps_popcnt,
        .fast_15bytenop,
        .fast_scalar_fsqrt,
        .fast_shld_rotate,
        .fast_variable_crosslane_shuffle,
        .fast_variable_perlane_shuffle,
        .fma,
        .fsgsbase,
        .fxsr,
        .idivq_to_divl,
        .invpcid,
        .lzcnt,
        .macrofusion,
        .mmx,
        .movbe,
        .nopl,
        .pclmul,
        .popcnt,
        .rdrnd,
        .sahf,
        .slow_3ops_lea,
        .sse,
        .sse2,
        .sse3,
        .sse4_1,
        .sse4_2,
        .ssse3,
        .vzeroupper,
        .x87,
        .xsave,
        .xsaveopt,
    }),
};
pub const os = std.Target.Os{
    .tag = .macos,
    .version_range = .{ .semver = .{
        .min = .{
            .major = 11,
            .minor = 6,
            .patch = 1,
        },
        .max = .{
            .major = 11,
            .minor = 6,
            .patch = 1,
        },
    }},
};
pub const target = std.Target{
    .cpu = cpu,
    .os = os,
    .abi = abi,
    .ofmt = object_format,
};
pub const object_format = std.Target.ObjectFormat.macho;
pub const mode = std.builtin.Mode.Debug;
pub const link_libc = true;
pub const link_libcpp = false;
pub const have_error_return_tracing = true;
pub const valgrind_support = false;
pub const sanitize_thread = false;
pub const position_independent_code = true;
pub const position_independent_executable = true;
pub const strip_debug_info = false;
pub const code_model = std.builtin.CodeModel.default;
```

**Comment by mikdusan** [comment]
> Output:
> 
> ```
> y=0
> ```
> 
> The LLVM IR emitted has `half 0xH0000` at the callsite. The compiler has incorrectly converted 4.0 to 0.

I am also able to reproduce this issue on:
- macos 13.2.1
- x86_64
- llvmorg-16.0.0 built using system toolchain
- and also zig+llvm tarball that CI failure used

**Comment by mikdusan** [comment]
AIR diff between master and llvm16 branch:
```diff
 # Begin Function AIR: bug0.main:
 # Total AIR+Liveness bytes: 503B
 # AIR Instructions:         27 (243B)
 # AIR Extra Data:           11 (44B)
 # AIR Values Bytes:         11 (88B)
 # Liveness tomb_bits:       16B
 # Liveness Extra Data:      0 (0B)
 # Liveness special table:   0 (0B)
   %0 = const_ty(*const type)
   %1 = constant(*const type, builtin)
   %2 = constant(type, builtin)
   %3 = const_ty(*const type)
   %4 = constant(*const type, builtin.StackTrace)
   %5 = constant(type, builtin.StackTrace)
   %6 = const_ty(builtin.StackTrace)
   %11 = const_ty(*const fn(f16) void)
   %12 = constant(*const fn(f16) void, (function 'foo'))
   %13 = const_ty(fn(f16) void)
   %14 = constant(fn(f16) void, (function 'foo'))
   %16 = constant(comptime_float, 4)
-  %17 = constant(f16, 4)
+  %17 = constant(f16, 0)
   %21 = const_ty(@typeInfo(@typeInfo(@TypeOf(bug0.main)).Fn.return_type.?).ErrorUnion.error_set!void)
   %22 = constant(@typeInfo(@typeInfo(@TypeOf(bug0.main)).Fn.return_type.?).ErrorUnion.error_set!void, {})
   %24 = const_ty(*const type)
   %25 = constant(*const type, builtin)
   %26 = constant(type, builtin)

   %7!= save_err_return_trace_index()
   %9!= dbg_block_begin()
   %10!= dbg_stmt(2:5)
   %15!= dbg_stmt(2:8)
   %18!= call(%14!, [%17!])
   %19!= dbg_block_end()
   %23!= ret(%22!)
 # End Function AIR: bug0.main
```

**Comment by mikdusan** [comment]
I created the following 2 executables:
- `zig3` regular build, llvm16, etc.
- `zig4` which steals the `zig ld` link command line from `zig3`, and applied it to system linker `ld` instead

results:
- `zig3` → reproduces issue
- `zig4` → does NOT reproduce issue

It looks like this is a zld issue.

I also offer a little more reduction which only asks the compiler to convert comptime_float to f16 and look at AIR or LLVM-IR and the bugged value is `constant(f16, 0)` and `half 0xH0000`, respectively:

### bug0.zig
```zig
pub fn main() !void {
    var bar: f16 = 4.0;
    _ = bar;
}
```

```
# Begin Function AIR: bug0.main:
# Total AIR+Liveness bytes: 495B
# AIR Instructions:         27 (243B)
# AIR Extra Data:           13 (52B)
# AIR Values Bytes:         9 (72B)
# Liveness tomb_bits:       16B
# Liveness Extra Data:      0 (0B)
# Liveness special table:   0 (0B)
  %0 = const_ty(*const type)
  %1 = constant(*const type, builtin)
  %2 = constant(type, builtin)
  %3 = const_ty(*const type)
  %4 = constant(*const type, builtin.StackTrace)
  %5 = constant(type, builtin.StackTrace)
  %6 = const_ty(builtin.StackTrace)
  %12 = constant(comptime_float, 4)
  %13 = constant(f16, 0)
  %15 = const_ty(*f16)
  %21 = const_ty(@typeInfo(@typeInfo(@TypeOf(bug0.main)).Fn.return_type.?).ErrorUnion.error_set!void)
  %22 = constant(@typeInfo(@typeInfo(@TypeOf(bug0.main)).Fn.return_type.?).ErrorUnion.error_set!void, {})
  %24 = const_ty(*const type)
  %25 = constant(*const type, builtin)
  %26 = constant(type, builtin)

  %7!= save_err_return_trace_index()
  %9!= dbg_block_begin()
  %10!= dbg_stmt(2:5)
  %11 = alloc(*f16)
  %14!= store(%11, %13!)
  %16!= dbg_var_ptr(%11, bar)
  %17!= dbg_stmt(3:5)
  %18!= load(f16, %11!)
  %19!= dbg_block_end()
  %23!= ret(%22!)
# End Function AIR: bug0.main
```

```llvm
; Function Attrs: nounwind sspstrong
define internal fastcc i16 @bug0.main(ptr nonnull %0) unnamed_addr #0 !dbg !1745 {
Entry:
  %1 = alloca half, align 2
  store half 0xH0000, ptr %1, align 2, !dbg !1754
  call void @llvm.dbg.declare(metadata ptr %1, metadata !1751, metadata !DIExpression()), !dbg !1754
  ret i16 0, !dbg !1755
}
```

**Comment by andrewrk** [comment]
@mikdusan you've already demonstrated that it's not a zld issue with your AIR diff above. The incorrect value 0 appears from semantic analysis long before the linker gets involved.

**Comment by kubkon** [comment]
> 

Ah good point!

**Comment by andrewrk** [comment]
No- I'm mistaken. zig2 does not exhibit the issue, and zig2 is linked with the system linker. Then zig3 is linked with zld, and the issue manifests.

However, the issue does not manifest with zld executed on an x86_64 Linux host to cross compile the macos binary.

**Comment by kubkon** [comment]
I'll take a look tomorrow and see if I can repro it locally, and try ruling out some things, etc.

**Comment by kubkon** [comment]
Also, for cross-compilation example, could it be that this feature is disabled/requires a specific MCPU value hence why it doesn't manifest itself in that scenario?

**Comment by andrewrk** [comment]
Yes that is likely. I believe LLVM 16 changed the ABI of f16 only when sse2 is enabled.

**Comment by mikdusan** [comment]
applying this diff and clean build of zig, issue does NOT manifest:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 326372501..e0b0c710c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -112,7 +112,7 @@ string(REGEX REPLACE "\\\\" "\\\\\\\\" ZIG_LIBC_STATIC_LIB_DIR_ESCAPED "${ZIG_LI
 string(REGEX REPLACE "\\\\" "\\\\\\\\" ZIG_LIBC_INCLUDE_DIR_ESCAPED "${ZIG_LIBC_INCLUDE_DIR}")
 
 set(ZIG_TARGET_TRIPLE "native" CACHE STRING "arch-os-abi to output binaries for")
-set(ZIG_TARGET_MCPU "native" CACHE STRING "-mcpu parameter to output binaries for")
+set(ZIG_TARGET_MCPU "baseline" CACHE STRING "-mcpu parameter to output binaries for")
 set(ZIG_SINGLE_THREADED off CACHE BOOL "limit the zig compiler to use only 1 thread")
 set(ZIG_AR_WORKAROUND off CACHE BOOL "append 'ar' subcommand to CMAKE_AR")
 ```

**Comment by mikdusan** [comment]
here's a reduction that at runtime casts `f64` to `f16`:

### `cast.zig`
```zig
const std = @import("std");

pub fn main() void {
    var num0: f64 = 4.0;
    var num1: f16 = 0.0;
    num1 = @floatCast(f16, num0);
    std.log.debug("{}", .{num1});
}
```

and in this shell session:
- build via `zig2` → executable `cast.zig2`
- build via `zig3` → executable `cast.zig3`
- re-link zig3 objects using system linker → executable `cast.zig3.ld`
- run the 3 executables...

```sh
$ rm -fr zig-cache; zig2 build-exe cast.zig -femit-bin=cast.zig2 --verbose-link
LLVM Emit Object... ar rcs /Users/mike/project/zig/work/llvm16/zig-cache/o/11180f8f15c683c4329fa75a0ac5e31c/libcompiler_rt.a /Users/mike/project/zig/work/llvm16/zig-cache/o/11180f8f15c683c4329fa75a0ac5e31c/libcompiler_rt.a.o
MachO Flush... zig ld -dynamic cast.zig2.o /Users/mike/project/zig/work/llvm16/zig-cache/o/11180f8f15c683c4329fa75a0ac5e31c/libcompiler_rt.a -o cast.zig2 -lSystem -lc

$ rm -fr zig-cache; zig3 build-exe cast.zig -femit-bin=cast.zig3 --verbose-link
LLVM Emit Object... ar rcs /Users/mike/project/zig/work/llvm16/zig-cache/o/850c20b177c37b631bc8d35e4a834396/libcompiler_rt.a /Users/mike/project/zig/work/llvm16/zig-cache/o/850c20b177c37b631bc8d35e4a834396/libcompiler_rt.a.o
MachO Flush... zig ld -dynamic cast.zig3.o /Users/mike/project/zig/work/llvm16/zig-cache/o/850c20b177c37b631bc8d35e4a834396/libcompiler_rt.a -o cast.zig3 -lSystem -lc

$ ld -dynamic cast.zig3.o /Users/mike/project/zig/work/llvm16/zig-cache/o/850c20b177c37b631bc8d35e4a834396/libcompiler_rt.a -o cast.zig3.ld -lSystem -L`xcrun --show-sdk-path`/usr/lib

$ ll cast.zig2 cast.zig3 cast.zig3.ld
-rwxr-xr-x 1 mike staff 903808 Mar 19 22:20 cast.zig2*
-rwxr-xr-x 1 mike staff 903808 Mar 19 22:21 cast.zig3*
-rwxr-xr-x 1 mike staff 954800 Mar 19 22:22 cast.zig3.ld*

$ ./cast.zig2
debug: 0.0e+00
# BAD

$ ./cast.zig3
debug: 0.0e+00
# BAD

$ ./cast.zig3.ld
debug: 4.0e+00
# GOOD
```

**Comment by jacobly0** [comment]
> it correctly printed y=4

Any reports of successful runs in this issue are extremely suspect because the above programs can all print the correct value *even with the miscompile*, it is all up to llvm regalloc ~~randomness~~ luck.  I suspect that there is a miscompile in every configuration tested, and offer the following, more reliable reproducer to confirm:

```zig
export fn f(x: f32) f16 {
    return @floatCast(f16, x);
}
```
```
$ zig build-obj f16.zig && objdump -d f16.o

f16.o:	file format mach-o 64-bit x86-64

Disassembly of section __TEXT,__text:

0000000000000000 <_f>:
       0: 55                           	pushq	%rbp
       1: 48 89 e5                     	movq	%rsp, %rbp
       4: 48 83 ec 10                  	subq	$0x10, %rsp
       8: f3 0f 11 45 fc               	movss	%xmm0, -0x4(%rbp)
       d: e8 00 00 00 00               	callq	0x12 <_f+0x12>
      12: 66 89 c1                     	movw	%ax, %cx
      15: 66 89 c8                     	movw	%cx, %ax
      18: 66 0f c4 c0 00               	pinsrw	$0x0, %eax, %xmm0
      1d: 48 83 c4 10                  	addq	$0x10, %rsp
      21: 5d                           	popq	%rbp
      22: c3                           	retq
```

If the instruction after `callq` is `movw %ax,...` then it is certainly miscompiled (at least relative to `compiler_rt.zig`) regardless of whether the value in that register happens to be correct.

**Comment by andrewrk** [comment]
fixed by 243848167b753692602609888763d47898ea4ce9 and 5870ffeb82e93ab2a35f9153668d170eced25cbc
